ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1627 @@
+
+    if (RAND_bytes_ex(libctx, cid->id, len, len * 8) != 1) {
+        cid->id_len = 0;
+        return 0;
+    }
+
+    return 1;
+}
+
+/*
+ * QUIC Channel Initialization and Teardown
+ * ========================================
+ */
+static int ch_init(QUIC_CHANNEL *ch)
+{
+    OSSL_QUIC_TX_PACKETISER_ARGS txp_args = {0};
+    OSSL_QTX_ARGS qtx_args = {0};
+    OSSL_QRX_ARGS qrx_args = {0};
+    QUIC_DHS_ARGS dhs_args = {0};
+    uint32_t pn_space;
+
+    /* TODO(QUIC): This is only applicable to clients. */
+    if (!gen_rand_conn_id(ch->libctx, INIT_DCID_LEN, &ch->init_dcid))
+        goto err;
+
+    /* We plug in a network write BIO to the QTX later when we get one. */
+    qtx_args.mdpl = QUIC_MIN_INITIAL_DGRAM_LEN;
+    ch->rx_max_udp_payload_size = qtx_args.mdpl;
+
+    ch->qtx = ossl_qtx_new(&qtx_args);
+    if (ch->qtx == NULL)
+        goto err;
+
+    ch->txpim = ossl_quic_txpim_new();
+    if (ch->txpim == NULL)
+        goto err;
+
+    ch->cfq = ossl_quic_cfq_new();
+    if (ch->cfq == NULL)
+        goto err;
+
+    if (!ossl_quic_txfc_init(&ch->conn_txfc, NULL))
+        goto err;
+
+    if (!ossl_quic_rxfc_init(&ch->conn_rxfc, NULL,
+                             2  * 1024 * 1024,
+                             10 * 1024 * 1024,
+                             get_time, NULL))
+        goto err;
+
+    if (!ossl_statm_init(&ch->statm))
+        goto err;
+
+    ch->have_statm = 1;
+    ch->cc_method = &ossl_cc_dummy_method;
+    if ((ch->cc_data = ch->cc_method->new(NULL, NULL, NULL)) == NULL)
+        goto err;
+
+    if ((ch->ackm = ossl_ackm_new(get_time, NULL, &ch->statm,
+                                  ch->cc_method, ch->cc_data)) == NULL)
+        goto err;
+
+    if (!ossl_quic_stream_map_init(&ch->qsm, get_stream_limit, ch))
+        goto err;
+
+    ch->have_qsm = 1;
+
+    /* We use a zero-length SCID. */
+    txp_args.cur_dcid           = ch->init_dcid;
+    txp_args.ack_delay_exponent = 3;
+    txp_args.qtx                = ch->qtx;
+    txp_args.txpim              = ch->txpim;
+    txp_args.cfq                = ch->cfq;
+    txp_args.ackm               = ch->ackm;
+    txp_args.qsm                = &ch->qsm;
+    txp_args.conn_txfc          = &ch->conn_txfc;
+    txp_args.conn_rxfc          = &ch->conn_rxfc;
+    txp_args.cc_method          = ch->cc_method;
+    txp_args.cc_data            = ch->cc_data;
+    txp_args.now                = get_time;
+    for (pn_space = QUIC_PN_SPACE_INITIAL; pn_space < QUIC_PN_SPACE_NUM; ++pn_space) {
+        ch->crypto_send[pn_space] = ossl_quic_sstream_new(INIT_CRYPTO_BUF_LEN);
+        if (ch->crypto_send[pn_space] == NULL)
+            goto err;
+
+        txp_args.crypto[pn_space] = ch->crypto_send[pn_space];
+    }
+
+    ch->txp = ossl_quic_tx_packetiser_new(&txp_args);
+    if (ch->txp == NULL)
+        goto err;
+
+    if ((ch->demux = ossl_quic_demux_new(/*BIO=*/NULL, /*Short CID Len=*/0,
+                                         1200, get_time, NULL)) == NULL)
```

Comment:
The use of 1200 here seems to be problematic. It ends up meaning we allocate a 1200 byte buffer for the incoming datagrams. If the datagram is longer than 1200 then it gets truncated.